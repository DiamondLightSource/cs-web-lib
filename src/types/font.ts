import { CSSProperties } from "react";

// assigned int values starting from zero
// also functions as a type
export enum FontStyle {
  Regular,
  Bold,
  Italic,
  BoldItalic
}

export class Font {
  // If no number is provided inherit from CSS.
  private size?: number;
  private style: FontStyle;
  private typeface: string;
  // TODO: name is currently unused
  private name?: string;

  public constructor(
    size?: number,
    style?: FontStyle,
    typeface?: string,
    name?: string
  ) {
    this.typeface = typeface ?? "Liberation sans";
    this.style = style ?? FontStyle.Regular;
    this.size = size;
    this.name = name;
  }

  /**
   * Returns a CSSProperties object containing all parameters defined by the
   * user and default properties for optional parameters not input by the user
   */
  public css(): CSSProperties {
    // Convert from px to REM by dividing pixels by default browser font size 16px
    // If a user has manually adjusted their browser font size, this might not be scaled
    // correctly, but given bob screens are pixel placed we shouldn't expect this to scale
    const fontSize = this.size ? `${this.size / 16}rem` : undefined;
    const fontWeight =
      this.style === FontStyle.Bold || this.style === FontStyle.BoldItalic
        ? "bold"
        : "normal";
    const fontStyle =
      this.style === FontStyle.Italic || this.style === FontStyle.BoldItalic
        ? "italic"
        : "normal";
    const style: CSSProperties = {
      // Fall back to sans-serif.
      fontFamily: `${this.typeface},sans-serif`,
      fontWeight: fontWeight,
      fontStyle: fontStyle
    };
    if (fontSize) {
      style.fontSize = fontSize;
    }
    return style;
  }
}
