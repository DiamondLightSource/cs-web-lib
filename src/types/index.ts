export { AbsolutePosition } from "./position";
export { RelativePosition } from "./position";
export { Color } from "./color";
export { PV } from "./pv";
export { Font, FontStyle } from "./font";
export { Border, BorderStyle } from "./border";
