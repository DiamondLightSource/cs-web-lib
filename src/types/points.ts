export class Point {
  public x: number;
  public y: number;

  public constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }
}

export class Points {
  public values: Array<Point>;

  public constructor(values: Array<Point>) {
    this.values = values;
  }
}
