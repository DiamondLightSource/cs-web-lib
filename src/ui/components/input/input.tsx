import { TextField, ThemeProvider } from "@mui/material";
import React, { CSSProperties, useState } from "react";
import { diamondTheme } from "../../../diamondTheme";

export const InputComponent = (props: {
  value: string;
  onEnter: (value: string) => void;
  readonly?: boolean;
  style?: CSSProperties;
  className?: string;
}): JSX.Element => {
  const [inputValue, setInputValue] = useState("");
  const [editing, setEditing] = useState(false);
  function onKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {
    if (event.key === "Enter") {
      props.onEnter(event.currentTarget.value);
      setInputValue("");
      setEditing(false);
      event.currentTarget.blur();
    }
  }

  function onClick(event: React.MouseEvent<HTMLInputElement>): void {
    /* When focus gained allow editing. */
    if (!props.readonly && !editing) {
      setInputValue("");
      setEditing(true);
    }
  }

  if (!editing && inputValue !== props.value) {
    setInputValue(props.value);
  }

  return (
    <ThemeProvider theme={diamondTheme}>
      <TextField
        variant="outlined"
        type="text"
        value={inputValue}
        slotProps={{ input: { readOnly: props.readonly } }}
        onKeyDown={onKeyDown}
        onChange={event => setInputValue(event.target.value)}
        onBlur={() => setEditing(false)}
        onClick={onClick}
        className={props.className}
        sx={{
          input: { color: props.style?.color ?? null }
        }}
      />
    </ThemeProvider>
  );
};
