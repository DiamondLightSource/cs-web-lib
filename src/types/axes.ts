import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public index: number;
  public autoScale?: boolean;
  public autoScaleThreshold?: number;
  public axisColor?: Color;
  public axisTitle?: string;
  public showGrid?: boolean;
  public gridColor?: Color;
  public dashGridLine?: boolean;
  public timeFormat?: number;
  public scaleFormat?: string;
  public scaleFont?: Font;
  public titleFont?: Font;
  public visible?: boolean;
  public logScale?: boolean;
  public leftBottomSide?: boolean;
  public maximum?: number;
  public minimum?: number;
  public yAxis?: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor(idx: number) {
    this.index = idx;
    this.autoScale = true;
    this.autoScaleThreshold = 0.95;
    this.axisColor = new Color("rgb(0, 0, 0");
    this.axisTitle = "";
    this.showGrid = false;
    this.gridColor = new Color("rgb(0, 0, 0");
    this.dashGridLine = true;
    this.timeFormat = 0;
    this.scaleFormat = "";
    this.scaleFont = new Font();
    this.titleFont = new Font(10, FontStyle.Bold);
    this.visible = true;
    this.logScale = false;
    this.leftBottomSide = true;
    this.minimum = 0;
    this.maximum = 100;
    // If axis has index 1 or above, assume y axis
    this.index > 0 ? (this.yAxis = true) : (this.yAxis = false);
  }
}

export class Axes {
  public count: number;
  public axisOptions: Axis[];

  public constructor(count: number, axes: Axis[]) {
    if (count !== axes.length) {
      throw new Error(
        `Count ${count} is not equal to number of axes ${axes.length}`
      );
    }
    this.count = count;
    this.axisOptions = axes;
  }
}
