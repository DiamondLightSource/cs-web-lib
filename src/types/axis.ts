import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public color: Color;
  public title: string;
  public showGrid: boolean;
  public visible: boolean;
  public logScale: boolean;
  public autoscale: boolean;
  public maximum: number;
  public minimum: number;
  public titleFont: Font;
  public scaleFont: Font;
  public onRight: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor({
    xAxis = false,
    color = new Color("rgb(0, 0, 0"),
    title = "",
    showGrid = false,
    visible = true,
    logScale = false,
    autoscale = false,
    minimum = 0,
    maximum = 100,
    titleFont = new Font(FontStyle.Bold),
    scaleFont = new Font(),
    onRight = false
  } = {}) {
    this.color = color;
    this.title = title || (xAxis ? "X" : "Y");
    this.showGrid = showGrid;
    this.visible = visible;
    this.logScale = logScale;
    this.autoscale = autoscale;
    this.minimum = minimum;
    this.maximum = maximum;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.onRight = onRight;
  }
}

export type Axes = Axis[];
