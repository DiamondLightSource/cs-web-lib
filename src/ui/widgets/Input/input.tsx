import React, { useEffect, useState } from "react";

import { writePv } from "../../hooks/useSubscription";
import { Widget } from "../widget";
import { PVInputComponent, PVWidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import {
  InferWidgetProps,
  FontPropOpt,
  ChoicePropOpt,
  ColorPropOpt,
  BoolPropOpt,
  BorderPropOpt,
  StringPropOpt,
  FloatPropOpt
} from "../propTypes";
import { AlarmQuality, DType } from "../../../types/dtypes";
import { TextField as MuiTextField, styled, useTheme } from "@mui/material";
import { WIDGET_DEFAULT_SIZES } from "../EmbeddedDisplay/bobParser";

const InputComponentProps = {
  pvName: StringPropOpt,
  font: FontPropOpt,
  foregroundColor: ColorPropOpt,
  backgroundColor: ColorPropOpt,
  transparent: BoolPropOpt,
  alarmSensitive: BoolPropOpt,
  enabled: BoolPropOpt,
  textAlign: ChoicePropOpt(["left", "center", "right"]),
  textAlignV: ChoicePropOpt(["top", "center", "bottom"]),
  border: BorderPropOpt,
  multiLine: BoolPropOpt,
  height: FloatPropOpt
};

const TextField = styled(MuiTextField)({
  // MUI Textfield contains a fieldset with a legend that needs to be removed
  "& .css-w4cd9x": {
    lineHeight: "0px"
  },
  "&.MuiFormControl-root": {
    height: "100%",
    width: "100%",
    display: "flex"
  },
  "& .MuiInputBase-root": {
    height: "100%",
    width: "100%",
    padding: "0px 4px"
  },
  "& .MuiInputBase-input": {
    padding: "0px",
    lineHeight: 1,
    textOverflow: "ellipsis",
    whiteSpace: "pre-wrap",
    height: "100%",
    width: "100%"
  },
  "& .MuiOutlinedInput-root": {
    "& .MuiOutlinedInput-notchedOutline": {
      borderRadius: "4px",
      borderWidth: "0px",
      inset: "0px"
    },
    "&.Mui-disabled": {
      cursor: "not-allowed",
      pointerEvents: "all !important"
    }
  }
});

export const SmartInputComponent = (
  props: PVInputComponent & InferWidgetProps<typeof InputComponentProps>
): JSX.Element => {
  const theme = useTheme();
  const {
    enabled = true,
    transparent = false,
    textAlign = "left",
    textAlignV = "center",
    value = null,
    multiLine = false,
    alarmSensitive = true,
    height = WIDGET_DEFAULT_SIZES["textupdate"][1]
  } = props;

  const font = props.font?.css() ?? theme.typography;

  let foregroundColor =
    props.foregroundColor?.toString() ?? theme.palette.primary.contrastText;

  let borderColor = props.border?.css().borderColor ?? "#000000";
  let borderStyle = props.border?.css().borderStyle ?? "solid";
  let borderWidth = props.border?.width ?? "0px";

  const alarmQuality = props.value?.getAlarm().quality ?? AlarmQuality.VALID;
  if (alarmSensitive) {
    switch (alarmQuality) {
      case AlarmQuality.UNDEFINED:
      case AlarmQuality.INVALID:
      case AlarmQuality.CHANGING:
        foregroundColor = "var(--invalid)";
        borderColor = "var(--invalid)";
        borderStyle = "solid";
        borderWidth = "1px";
        break;
      case AlarmQuality.ALARM:
      case AlarmQuality.WARNING:
        foregroundColor = "var(--alarm)";
        borderColor = "var(--alarm)";
        borderStyle = "solid";
        borderWidth = "2px";
        break;
      default:
        break;
    }
  }

  let alignmentV = "center";
  if (textAlignV === "top") {
    alignmentV = "start";
  } else if (textAlignV === "bottom") {
    alignmentV = "end";
  }

  const backgroundColor = transparent
    ? "transparent"
    : (props.backgroundColor?.toString() ?? "#80FFFF");

  // If props.font exists, extracts the font size in rem and returns is back to size in px
  // using the default browser size of 16px, as used in ../../../types/font.ts
  const fontSize = props.font?.css().fontSize
    ? parseFloat(
        props.font
          .css()
          .fontSize?.toString()
          .match(/\d+.\d+/)
          ?.toString() ?? ""
      ) * 16
    : theme.typography.fontSize;

  const maxRows = multiLine
    ? Math.floor(height / fontSize) - 1 < 1
      ? 1
      : Math.floor(height / fontSize) - 1
    : 1;

  const [inputValue, setInputValue] = useState(value?.getStringValue() ?? "");

  useEffect(() => {
    if (value) {
      setInputValue(value.getStringValue() ?? "");
    }
  }, [value]);

  const onKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (multiLine) {
      if (event.key === "Enter" && event.ctrlKey) {
        writePv(props.pvName, new DType({ stringValue: inputValue }));
        event.currentTarget.blur();
      }
    } else {
      if (event.key === "Enter") {
        writePv(props.pvName, new DType({ stringValue: inputValue }));
        event.currentTarget.blur();
      }
    }
  };

  return (
    <TextField
      disabled={!enabled}
      value={inputValue}
      maxRows={maxRows}
      multiline={multiLine}
      variant="outlined"
      type="text"
      slotProps={{
        input: {
          onKeyDown: onKeyPress
        }
      }}
      onChange={event => setInputValue(event.target.value)}
      sx={{
        "& .MuiInputBase-input": {
          textAlign: textAlign,
          font: font
        },
        "& .MuiInputBase-root": {
          alignItems: alignmentV,
          color: foregroundColor,
          backgroundColor: backgroundColor
        },
        "& .MuiOutlinedInput-root": {
          "& .MuiOutlinedInput-notchedOutline": {
            outlineWidth: borderWidth,
            outlineStyle: borderStyle,
            outlineColor: borderColor
          },
          "&.Mui-focused": {
            "& .MuiOutlinedInput-notchedOutline": {
              borderWidth: "0px"
            }
          }
        }
      }}
    />
  );
};

const InputWidgetProps = {
  ...InputComponentProps,
  ...PVWidgetPropType
};

export const Input = (
  props: InferWidgetProps<typeof InputWidgetProps>
): JSX.Element => <Widget baseWidget={SmartInputComponent} {...props} />;

registerWidget(Input, InputWidgetProps, "input");
