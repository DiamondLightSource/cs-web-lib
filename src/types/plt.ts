import { Axes, Axis } from "./axis";
import { Color } from "./color";
import { Font } from "./font";
import { Trace } from "./trace";

export class Plt {
  public title: string;
  public axes: Axes;
  public pvlist: Trace[];
  public backgroundColor: Color;
  public foregroundColor: Color;
  public scroll: boolean;
  public scrollStep: number;
  public updatePeriod: number;
  public start: string;
  public end: string;
  public showGrid: boolean;
  public titleFont: Font;
  public scaleFont: Font;
  public labelFont: Font;
  public legendFont: Font;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property.
   */
  public constructor({
    title = "",
    axes = [new Axis()],
    pvlist = [new Trace()],
    background = Color.WHITE,
    foreground = Color.BLACK,
    scroll = true,
    grid = false,
    scrollStep = 5,
    updatePeriod = 1,
    titleFont = new Font(),
    labelFont = new Font(),
    legendFont = new Font(),
    scaleFont = new Font(),
    start = "1 minute",
    end = "now"
  } = {}) {
    this.backgroundColor = background;
    this.foregroundColor = foreground;
    this.title = title;
    this.scroll = scroll;
    this.scrollStep = scrollStep;
    this.axes = axes;
    this.pvlist = pvlist;
    this.updatePeriod = updatePeriod;
    this.showGrid = grid;
    this.start = start;
    this.end = end;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.labelFont = labelFont;
    this.legendFont = legendFont;
  }
}
