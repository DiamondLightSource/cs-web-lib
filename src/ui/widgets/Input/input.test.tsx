import React from "react";
import { SmartInputComponent } from "./input";
import { render } from "@testing-library/react";
import { DAlarm } from "../../../types/dtypes";
import { dstring } from "../../../testResources";
import { PV, RelativePosition } from "../../../types";

let input: JSX.Element;
const pv = new PV("loc://test", "ca");
beforeEach((): void => {
  input = (
    <SmartInputComponent
      pvName={pv}
      position={new RelativePosition()}
      value={dstring("hello", DAlarm.MINOR)}
      connected={true}
      readonly={true}
      alarmSensitive={true}
    />
  );
});
describe("<Input />", (): void => {
  it("renders an input", (): void => {
    const { asFragment } = render(input);
    expect(asFragment()).toMatchSnapshot();
  });
});
