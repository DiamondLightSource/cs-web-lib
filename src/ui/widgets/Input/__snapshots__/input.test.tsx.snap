// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Input /> > renders an input 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root css-1a8n6hf-MuiFormControl-root-MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r0:"
        type="text"
        value="hello"
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-w4cd9x"
        >
          <span
            class="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
